core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/provision/service/StackCreationService.java

```diff
@@ -279,14 +284,45 @@ public void setInstanceStoreCount(StackContext stackContext) {
         InstanceStoreMetadata instanceStoreMetadata = connector.metadata().collectInstanceStorageCount(ac, instanceTypes);
 
         for (InstanceGroup ig : stack.getInstanceGroups()) {
-            if (ig.getTemplate() != null) {
-                ig.getTemplate().setInstanceStorageCount(
-                        instanceStoreMetadata.mapInstanceTypeToInstanceStoreCountNullHandled(ig.getTemplate().getInstanceType()));
-                templateService.savePure(ig.getTemplate());
+            Template template = ig.getTemplate();
+            if (template != null) {
+                addInstanceStorageVolumeTemplateToTemplate(instanceStoreMetadata, ig, template);
+                Integer instanceStorageCount = instanceStoreMetadata.mapInstanceTypeToInstanceStoreCountNullHandled(template.getInstanceType());
+                LOGGER.debug("Setting instance storage count in template. " +
+                        "Group name: {}, Template id: {}, instance type: {}", ig.getGroupName(), template.getId(), template.getInstanceType());
+                template.setInstanceStorageCount(instanceStorageCount);
+                templateService.savePure(template);
             }
         }
     }
 
+    private void addInstanceStorageVolumeTemplateToTemplate(InstanceStoreMetadata instanceStoreMetadata, InstanceGroup ig, Template template) {
+        if (!isTemplateSetUpForInstanceStorage(template)) {
+            LOGGER.debug("Instance storage was not requested. Attempting to add them implicitly. " +
+                    "Group name: {}, instance type: {}", ig.getGroupName(), template.getInstanceType());
+            Optional.ofNullable(instanceStoreMetadata.getInstaceStoreConfigMap().get(template.getInstanceType()))
```

Comment:
If `getInstaceStoreConfigMap()` could not be null I am fine with this call, but otherwise null check would needed here. 😸

Reply:
can you please also cover this logic with unit tests?